modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/IgniteH2Indexing.java

```diff
@@ -1104,7 +1150,31 @@ public void bindParameters(PreparedStatement stmt, @Nullable Collection<Object>
 
         twoStepQry.pageSize(qry.getPageSize());
 
-        QueryCursorImpl<List<?>> cursor = new QueryCursorImpl<>(queryTwoStep(cctx, twoStepQry, cctx.keepBinary()));
+        final AtomicReference<GridAbsClosure> mapQrysCancel = new AtomicReference<>();
+        final AtomicReference<GridAbsClosure> rdcQryCancel = new AtomicReference<>();
+
+        QueryCursorImpl<List<?>> cursor = new QueryCursorImpl<>(
+            doQueryTwoStep(cctx, twoStepQry, cctx.keepBinary(), qry.getTimeout(), mapQrysCancel, rdcQryCancel),
+            new GridAbsClosure() {
+                @Override public void apply() {
+                    GridAbsClosure clo;
+                        // Cancellation closure is changed on each attempt. Make sure we use correct closure.
+                        while(true) {
+                            clo = mapQrysCancel.get();
+
+                            if (mapQrysCancel.compareAndSet(clo, F.noop()))
+                                break;
+                        }
+
+                        if (clo != null)
+                            clo.apply();
+
+                        clo = rdcQryCancel.get();
+
+                        if (clo != null && rdcQryCancel.compareAndSet(clo, F.noop()))
```

Comment:
This code snippet I see multiple times. It must be moved to static utility method.